src/tactic/sine_filter.cpp

```diff
@@ -0,0 +1,252 @@
+/*++
+Copyright (c) 2016 Microsoft Corporation
+
+Module Name:
+
+    sine_filter.cpp
+
+Abstract:
+
+    Tactic that performs Sine Qua Non premise selection
+
+Revision History:
+--*/
+
+#include "sine_filter.h"
+#include "tactical.h"
+#include "filter_model_converter.h"
+#include "datatype_decl_plugin.h"
+#include "rewriter_def.h"
+#include "filter_model_converter.h"
+#include "extension_model_converter.h"
+#include "var_subst.h"
+#include "ast_util.h"
+#include "obj_pair_hashtable.h"
+
+#if 0
+TODO documentation here
+#endif
+
+
+class sine_tactic : public tactic {
+
+    ast_manager&  m;
+    params_ref    m_params;
+    datatype_util m_dt;
+    ref<extension_model_converter> m_ext;
+    ref<filter_model_converter>    m_filter;
+    unsigned                       m_num_transformed;
+
+public:
+
+    sine_tactic(ast_manager& m, params_ref const& p): 
+        m(m), m_params(p), m_dt(m) {}
+    
+    virtual tactic * translate(ast_manager & m) {
+        return alloc(sine_tactic, m, m_params);
+    }
+
+    virtual void updt_params(params_ref const & p) {
+    }
+
+    virtual void collect_param_descrs(param_descrs & r) {
+    }
+
+    virtual void operator()(goal_ref const & g,
+                            goal_ref_buffer & result,
+                            model_converter_ref & mc,
+                            proof_converter_ref & pc,
+                            expr_dependency_ref & core) {
+        mc = 0; pc = 0; core = 0;
+
+        TRACE("sine", g->display(tout););
+        TRACE("sine", tout << g->size(););
+        ptr_vector<expr> new_forms;
+        filter_expressions(g, new_forms);
+        TRACE("sine", tout << new_forms.size(););
+        g->reset();
+        for (unsigned i = 0; i < new_forms.size(); i++) {
+          g->assert_expr(new_forms.get(i), 0, 0);
+        }
+        g->inc_depth();
+        g->updt_prec(goal::OVER);
+        result.push_back(g.get());
+        TRACE("sine", result[0]->display(tout););
+        SASSERT(g->is_well_sorted());
+        filter_model_converter * fmc = alloc(filter_model_converter, m);
+        mc = fmc;
+    }
+    
+    virtual void cleanup() {
+    }
+
+private:
+
+  // is this a user-defined symbol name?
+  bool is_name(func_decl * f) {
+    return f->get_family_id() < 0;
```

Comment:
There must be some better method than this. What is the specification of this function? We have methods called "is_uninterp" or uninterp_const or something like this for user-defined functions.